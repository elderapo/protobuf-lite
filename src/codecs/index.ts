export * from "./DateCodec";
export * from "./ICustomFieldCodec";
export * from "./InternalCodecs";
