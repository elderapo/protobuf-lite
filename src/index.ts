export * from "./collectSchema";
export * from "./encoderDecoderFunctions";
export * from "./ProtobufLiteProperty";
