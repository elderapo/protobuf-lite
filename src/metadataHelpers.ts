import { ProtobufLiteMetadata } from "./ProtobufLiteMetadata";
import { getPrototypeChain } from "./utils";

// dont use Reflect.defineMetadata because it's slow AF
let weakMap = new WeakMap<Function, ProtobufLiteMetadata>();

export const clearAllMetadata = () => {
  weakMap = new WeakMap();
};

export const getMetadataObject = (Class: Function): ProtobufLiteMetadata => {
  if (!weakMap.has(Class)) {
    weakMap.set(Class, new ProtobufLiteMetadata(Class));
  }

  return weakMap.get(Class) as ProtobufLiteMetadata;
};

export const hasMetadataObject = (Class: Function): boolean => {
  const chain = getPrototypeChain(Class);

  for (let prototype of chain) {
    if (weakMap.has(prototype.constructor)) {
      return true;
    }
  }

  return false;
};
