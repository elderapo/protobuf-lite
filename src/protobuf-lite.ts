import "buffer/"; // polyfil for browser

export * from "./encoderDecoderFunctionts";
export * from "./ProtobufLiteProperty";
