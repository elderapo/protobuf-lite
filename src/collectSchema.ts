import { getMetadataObject, hasMetadataObject } from "./metadataHelpers";
import { IFieldInfo, IFieldOptions, ProtobufLiteMetadata } from "./ProtobufLiteMetadata";
import { getPrototypeChain, Omit } from "./utils";

export type FieldInfoWithoutPropertyKey = Omit<IFieldInfo, "propertyKey">;

export interface ISchema {
  refs: FieldInfoWithoutPropertyKey[][];
  fieldsInfo: FieldInfoWithoutPropertyKey[];
}

export const collectSchema = (MessageClass: Function): ISchema => {
  /* istanbul ignore next line */
  if (!hasMetadataObject(MessageClass)) {
    throw new Error(`MessageClass doesn't have protobuf lite metadata assosiated!`);
  }

  const metadata = getMetadataObject(MessageClass);

  const schema: ISchema = {
    refs: [],
    fieldsInfo: []
  };

  const mapField = (fieldInfo: IFieldInfo): FieldInfoWithoutPropertyKey => {
    if (childTypesByClassName.has(fieldInfo.prototype)) {
      const fieldOptions = childTypesByClassName.get(fieldInfo.prototype) as IFieldOptions;
      const indexRef = indexedChildTypes.indexOf(fieldOptions);

      return { prototype: `__ref__${indexRef}__`, rule: fieldInfo.rule };
    }

    return { prototype: fieldInfo.prototype, rule: fieldInfo.rule };
  };

  const childTypesByClassName = new Map<string, IFieldOptions>();
  const indexedChildTypes: IFieldOptions[] = [];

  const collectChildTypes = (childTypes: IFieldOptions[]) => {
    for (let item of childTypes) {
      const metadata = getMetadataObject(item.MessageClass);

      if (!childTypesByClassName.has(metadata.getMessageClassName())) {
        childTypesByClassName.set(metadata.getMessageClassName(), item);
        indexedChildTypes.push(item);

        if (metadata.getChildTypes()) {
          collectChildTypes(metadata.getChildTypes());
        }
      }
    }
  };

  collectChildTypes(metadata.getChildTypes());

  const alreadyCollectedChildTypes = new Map<ProtobufLiteMetadata, boolean>();

  const handleChildTypes = (childTypes: IFieldOptions[]) => {
    for (let item of childTypes) {
      const metadata = getMetadataObject(item.MessageClass);

      if (alreadyCollectedChildTypes.has(metadata)) {
        continue;
      }
      alreadyCollectedChildTypes.set(metadata, true);

      schema.refs.push(metadata.getFieldsInfo().map(mapField));

      if (metadata.getChildTypes()) {
        handleChildTypes(metadata.getChildTypes());
      }
    }
  };

  handleChildTypes(metadata.getChildTypes());

  // const MessageClass = metadata.getMessageClass();
  const prototypes = getPrototypeChain(MessageClass)
    .reverse()
    .filter(p => p !== MessageClass.prototype);

  const alreadyUsedPropertyKeys: Map<string, boolean> = new Map();

  const addFields = (fields: IFieldInfo[]) => {
    for (let field of fields) {
      if (alreadyUsedPropertyKeys.has(field.propertyKey)) {
        throw new Error(`Parent class field was most likely overwrited by child class!`);
      }

      schema.fieldsInfo.push(mapField(field));

      alreadyUsedPropertyKeys.set(field.propertyKey, true);
    }
  };

  for (let prototype of prototypes) {
    const mt = getMetadataObject(prototype.constructor);
    const fields = mt.getFieldsInfo();

    addFields(fields);
  }

  addFields(metadata.getFieldsInfo());

  return schema;
};
