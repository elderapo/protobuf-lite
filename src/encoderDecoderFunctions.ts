import { Message } from "protobufjs/light";
import { getMetadataObject, hasMetadataObject } from "./metadataHelpers";
import { ensureBuffer, Constructable } from "./utils";

export const encode = <T extends Object>(MessageClass: Constructable<T>, payload: T): Buffer => {
  if (!payload) {
    throw new Error(`Payload wasnt provided!`);
  }

  if (!hasMetadataObject(MessageClass)) {
    throw new Error(`MessageClass doesn't have protobuf lite metadata assosiated!`);
  }

  const metadataObject = getMetadataObject(MessageClass);
  const MessageClassProto = metadataObject.getProto();

  payload = metadataObject.runCustomEncoders(payload);

  const errMsg = MessageClassProto.verify(payload);

  if (errMsg) {
    throw new Error(errMsg);
  }

  return ensureBuffer(MessageClassProto.encode(payload).finish());
};

export const decode = <T extends Object>(MessageClass: Constructable<T>, encoded: Buffer): T => {
  if (!hasMetadataObject(MessageClass)) {
    throw new Error(`MessageClass doesn't have protobuf lite metadata assosiated!`);
  }

  const metadataObject = getMetadataObject(MessageClass);
  const MessageClassProto = metadataObject.getProto();

  const decoded: Message<T> = MessageClassProto.decode(encoded);

  metadataObject.runCustomDecoders(decoded);
  metadataObject.fixPrototypes(decoded);

  return (decoded as unknown) as T;
};
