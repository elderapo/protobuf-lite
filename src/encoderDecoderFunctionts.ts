import { Message } from "protobufjs/light";
import { getMetadataObject, hasMetadataObject } from "./metadataHelpers";
import { ensureBuffer } from "./utils";

export const encode = <T>(MessageClass: new () => T, payload: T): Buffer => {
  if (!payload) {
    throw new Error(`Payload wasnt provided!`);
  }

  if (!hasMetadataObject(MessageClass)) {
    throw new Error(`MessageClass doesn't have protobuf lite metadata assosiated!`);
  }

  const metadataObject = getMetadataObject(MessageClass);
  const MessageClassProto = metadataObject.getProto();

  payload = metadataObject.runCustomEncoders(payload);

  const errMsg = MessageClassProto.verify(payload);

  if (errMsg) {
    throw new Error(errMsg);
  }

  return ensureBuffer(MessageClassProto.encode(payload).finish());
};

export const decode = <T extends Object>(MessageClass: new () => T, encoded: Buffer): T => {
  if (!hasMetadataObject(MessageClass)) {
    throw new Error(`MessageClass doesn't have protobuf lite metadata assosiated!`);
  }

  const metadataObject = getMetadataObject(MessageClass);
  const MessageClassProto = metadataObject.getProto();

  const decoded: Message<T> = MessageClassProto.decode(encoded);

  metadataObject.runCustomDecoders(decoded);
  metadataObject.fixPrototypes(decoded);

  return (decoded as unknown) as T;
};
