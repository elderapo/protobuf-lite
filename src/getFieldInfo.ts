import { getMetadataObject, hasMetadataObject } from "./metadataHelpers";
import { IFieldInfo } from "./ProtobufLiteMetadata";
import { Constructable } from "./utils";

export const getFieldInfo = (Class: Constructable<Object>): IFieldInfo[] => {
  if (!hasMetadataObject(Class)) {
    throw new Error(
      `Class(${Class}) does not have metadata assosiated with it. It's not possible to calculate checksum!`
    );
  }

  const metadataObject = getMetadataObject(Class);

  return metadataObject.collectFieldsInfo();
};
