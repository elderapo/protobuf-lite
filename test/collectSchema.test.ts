import "@abraham/reflection";
import { ProtobufLiteProperty } from "../src";
import {} from "../src/metadataHelpers";
import { collectSchema } from "../src/collectSchema";

describe("collectSchema(MessageClass)", () => {
  it("should return correct schema for strings", () => {
    class C1 {
      @ProtobufLiteProperty()
      name: string;
    }

    class C2 {
      @ProtobufLiteProperty({ optional: true })
      name?: string;
    }

    class C3 {
      @ProtobufLiteProperty({ type: () => String })
      names: string[];
    }

    expect(collectSchema(C1)).toMatchSnapshot();

    expect(collectSchema(C2)).toMatchSnapshot();

    expect(collectSchema(C3)).toMatchSnapshot();
  });

  it("should return correct schema for numbers", () => {
    class C1 {
      @ProtobufLiteProperty()
      age: number;
    }

    class C2 {
      @ProtobufLiteProperty({ optional: true })
      age?: number;
    }

    class C3 {
      @ProtobufLiteProperty({ type: () => Number })
      ages: number[];
    }

    expect(collectSchema(C1)).toMatchSnapshot();

    expect(collectSchema(C2)).toMatchSnapshot();

    expect(collectSchema(C3)).toMatchSnapshot();
  });

  it("should return correct schema for booleans", () => {
    class C1 {
      @ProtobufLiteProperty()
      isTrue: boolean;
    }

    class C2 {
      @ProtobufLiteProperty({ optional: true })
      isTrue?: boolean;
    }

    class C3 {
      @ProtobufLiteProperty({ type: () => Boolean })
      isTrues: boolean[];
    }

    expect(collectSchema(C1)).toMatchSnapshot();

    expect(collectSchema(C2)).toMatchSnapshot();

    expect(collectSchema(C3)).toMatchSnapshot();
  });

  it("should return correct schema for Buffers", () => {
    class C1 {
      @ProtobufLiteProperty()
      buffer: Buffer;
    }

    class C2 {
      @ProtobufLiteProperty({ optional: true })
      buffer?: Buffer;
    }

    class C3 {
      @ProtobufLiteProperty({ type: () => Buffer })
      buffers: Buffer[];
    }

    expect(collectSchema(C1)).toMatchSnapshot();

    expect(collectSchema(C2)).toMatchSnapshot();

    expect(collectSchema(C3)).toMatchSnapshot();
  });

  it("should return correct schema for Dates", () => {
    class C1 {
      @ProtobufLiteProperty()
      date: Date;
    }

    class C2 {
      @ProtobufLiteProperty({ optional: true })
      date?: Date;
    }

    class C3 {
      @ProtobufLiteProperty({ type: () => Date })
      dates: Date[];
    }

    expect(collectSchema(C1)).toMatchSnapshot();

    expect(collectSchema(C2)).toMatchSnapshot();

    expect(collectSchema(C3)).toMatchSnapshot();
  });

  it("should return correct schema for inhreited classes v1", () => {
    class Parent {
      @ProtobufLiteProperty()
      someField: string;
    }

    class Child extends Parent {
      @ProtobufLiteProperty()
      someOtherField: string;
    }

    expect(collectSchema(Child)).toMatchSnapshot();
  });

  it("should return correct schema for inhreited classes v2", () => {
    class C1 {
      @ProtobufLiteProperty()
      c1: string;
    }

    class C2 extends C1 {
      @ProtobufLiteProperty()
      c2: string;
    }

    class C3 extends C2 {
      @ProtobufLiteProperty()
      c3: string;
    }

    class C4 extends C3 {
      @ProtobufLiteProperty()
      c4: string;
    }

    class C5 extends C4 {
      @ProtobufLiteProperty()
      c5: string;
    }

    expect(collectSchema(C1)).toMatchSnapshot();

    expect(collectSchema(C2)).toMatchSnapshot();

    expect(collectSchema(C3)).toMatchSnapshot();

    expect(collectSchema(C4)).toMatchSnapshot();

    expect(collectSchema(C5)).toMatchSnapshot();
  });

  it("should return correct schema for inhreited classes v3", () => {
    class Parent {
      @ProtobufLiteProperty()
      someField: string;
    }

    class Child extends Parent {}
    class Child1 extends Child {}
    class Child2 extends Child1 {}
    class Child3 extends Child1 {
      @ProtobufLiteProperty()
      c3: string;
    }

    expect(collectSchema(Child)).toMatchSnapshot();

    expect(collectSchema(Child1)).toMatchSnapshot();

    expect(collectSchema(Child2)).toMatchSnapshot();

    expect(collectSchema(Child3)).toMatchSnapshot();
  });

  it("should return correct schema for nestedClasses", () => {
    class C3 {
      @ProtobufLiteProperty()
      iAmC3: string;
    }

    class C2 {
      @ProtobufLiteProperty()
      c3: C3;

      @ProtobufLiteProperty()
      iAmC2: string;
    }

    class C1 {
      @ProtobufLiteProperty()
      c2: C2;

      @ProtobufLiteProperty()
      iAmC1: string;
    }

    class CMain {
      @ProtobufLiteProperty()
      c1_0: C1;

      @ProtobufLiteProperty()
      c1_1: C1;

      @ProtobufLiteProperty()
      c1_2: C1;

      @ProtobufLiteProperty()
      cMainYooo: string;
    }

    expect(collectSchema(CMain)).toMatchSnapshot();
  });

  it("should throw if child overwrites parent fields", () => {
    class Parent {
      @ProtobufLiteProperty()
      someField: string;
    }

    class Child extends Parent {
      @ProtobufLiteProperty()
      someField: string;
    }

    expect(() => collectSchema(Child)).toThrowError();
  });
});
