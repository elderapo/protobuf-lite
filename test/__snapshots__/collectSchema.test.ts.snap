// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectSchema(MessageClass) should return correct schema for Buffers 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bytes",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for Buffers 2`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bytes",
      "rule": "optional",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for Buffers 3`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bytes",
      "rule": "repeated",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for Dates 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bytes",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for Dates 2`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bytes",
      "rule": "optional",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for Dates 3`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bytes",
      "rule": "repeated",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for booleans 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bool",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for booleans 2`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bool",
      "rule": "optional",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for booleans 3`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "bool",
      "rule": "repeated",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v1 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v2 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v2 2`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v2 3`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v2 4`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v2 5`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v3 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v3 2`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v3 3`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for inhreited classes v3 4`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for nestedClasses 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "__ref__0__",
      "rule": "required",
    },
    Object {
      "prototype": "__ref__0__",
      "rule": "required",
    },
    Object {
      "prototype": "__ref__0__",
      "rule": "required",
    },
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [
    Array [
      Object {
        "prototype": "__ref__1__",
        "rule": "required",
      },
      Object {
        "prototype": "string",
        "rule": "required",
      },
    ],
    Array [
      Object {
        "prototype": "__ref__2__",
        "rule": "required",
      },
      Object {
        "prototype": "string",
        "rule": "required",
      },
    ],
    Array [
      Object {
        "prototype": "string",
        "rule": "required",
      },
    ],
  ],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for numbers 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "int32",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for numbers 2`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "int32",
      "rule": "optional",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for numbers 3`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "int32",
      "rule": "repeated",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for strings 1`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "required",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for strings 2`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "optional",
    },
  ],
  "refs": Array [],
}
`;

exports[`collectSchema(MessageClass) should return correct schema for strings 3`] = `
Object {
  "fieldsInfo": Array [
    Object {
      "prototype": "string",
      "rule": "repeated",
    },
  ],
  "refs": Array [],
}
`;
